lib/Frontend/CachingUtils.cpp

```diff
@@ -133,6 +133,134 @@ createSwiftCachingOutputBackend(
                                                     InputsAndOutputs);
 }
 
+bool replayCachedCompilerOutputs(
+    ObjectStore &CAS, ActionCache &Cache, ObjectRef BaseKey,
+    DiagnosticEngine &Diag, const FrontendInputsAndOutputs &InputsAndOutputs,
+    CachingDiagnosticsProcessor &CDP) {
+  clang::cas::CompileJobResultSchema Schema(CAS);
+  bool CanReplayAllOutput = true;
+  struct OutputEntry {
+    std::string Path;
+    std::string Key;
+    llvm::cas::ObjectProxy Proxy;
+  };
+  SmallVector<OutputEntry> OutputProxies;
+
+  auto replayOutputFile = [&](StringRef InputName, file_types::ID OutputKind,
+                              StringRef OutputPath) -> Optional<OutputEntry> {
+    LLVM_DEBUG(llvm::dbgs()
+                   << "DEBUG: lookup output \'" << OutputPath << "\' type \'"
+                   << file_types::getTypeName(OutputKind) << "\' input \'"
+                   << InputName << "\n";);
+
+    auto OutputKey =
+        createCompileJobCacheKeyForOutput(CAS, BaseKey, InputName, OutputKind);
+    if (!OutputKey) {
+      Diag.diagnose(SourceLoc(), diag::error_cas,
+                    toString(OutputKey.takeError()));
+      return None;
+    }
+    auto OutputKeyID = CAS.getID(*OutputKey);
+    auto Lookup = Cache.get(OutputKeyID);
+    if (!Lookup) {
+      Diag.diagnose(SourceLoc(), diag::error_cas, toString(Lookup.takeError()));
+      return None;
+    }
+    if (!*Lookup) {
+      Diag.diagnose(SourceLoc(), diag::output_cache_miss, OutputPath,
```

Comment:
Going forward, should these remarks be being a flag of some sort?

Reply:
Yeah, we should guard them but at some point I will reconfigure all the flags. Currently I am thinking that:
* add `-compile-cache` driver option to turn on replay and output storing.
* `-enable-cas` will just be a frontend option to add a cas instance for various cas tasks, like cas based dependency scanning, cas fs, cas object file format, etc.